Improve device and code compute capability check

lakshayg opened the pull request:
Here's what the revised warning looks like

```
torch/cuda/__init__.py:340: UserWarning: Found GPU0 ... which is of compute capability (CC) 9.0.
The following list shows the CCs this version of PyTorch was built for and the hardware CCs it supports:
- 7.5 which supports hardware CC >=7.5,<8.0
- 8.0 which supports hardware CC >=8.0,<9.0 except {8.7}
- 8.6 which supports hardware CC >=8.6,<9.0 except {8.7}
- 10.0 which supports hardware CC >=10.0,<11.0 except {10.1}
- 10.1 which supports hardware CC {10.1, 11.0}
Please follow the instructions at https://pytorch.org/get-started/locally/ to install a PyTorch release that supports one of these CUDA versions: 12.6, 12.8, 13.0
```

fixes: #164708

Reference: NVIDIA Compute Capability to Device Mapping: https://developer.nvidia.com/cuda-gpus 

cc: @ptrblck @nWEIdia @tinglvv

nWEIdia commented:
> @pytorchbot label "topic: not user facing"

I would argue this is `user facing` , as potentially this code path could produce something the user might not be seeing previously. Or it fixes something user was seeing (e.g. on Spark), which was the main motivation of this PR. The user won't see the warning anymore e.g. while running pytorch nightly cu13.0 binary with this PR in.

lakshayg commented:
> > @lakshayg: @pytorchbot label "topic: not user facing"
> 
> @nWEIdia: I would argue this is `user facing` , ...

I have added the "release notes: cuda" label. Is that the right one or do you recommend a different one?

nWEIdia reviewed (approved):
LGTM, thanks!

atalman reviewed (approved):
lgtm, do we have a link to Nvidia documentation where device and code compute capability is defined ? This way will be easier to update the code for future devices

nWEIdia commented:
Yes @atalman : I updated the PR description with this official link: https://developer.nvidia.com/cuda-gpus

lakshayg commented:
@nWEIdia I believe @ptrblck wanted to review it as well. He wanted to think through the case when someone is building pytorch with custom arch list.

lakshayg commented:
@pytorchmergebot rebase

atalman commented:
Hi @lakshayg is this ready to be landed ?

lakshayg commented:
@atalman it's ready IMO but @ptrblck might want to take a look

lakshayg commented on test/test_cuda_compatibility.py:
```diff
@@ -0,0 +1,154 @@
+            torch.cuda._check_capability()
+            self.assertEqual(len(w), 1)
+            msg = str(w[0].message)
+            self.assertIn("12.6", msg)
```
I have mocked the relevant dict so the test will not need to be changed

busFred commented:
i'm on spark with torch 2.9.1 cu13.0 and i'm also receiving the same error. i skim through this discussion chain, correct me if i'm wrong, it seems this warning is just a false alarm and does not affect actual usability; I guess I could just ignore this until the next update comes out(?)

ptrblck commented:
@busFred Yes, ignore the warning assuming you are already using any of our PyTorch builds with CUDA 13.x on Spark.

nWEIdia commented:
@pytorchbot merge

lakshayg committed:
Improve device and code compute capability check (#166185)

Here's what the revised warning looks like

```
torch/cuda/__init__.py:340: UserWarning: Found GPU0 ... which is of compute capability (CC) 9.0.
The following list shows the CCs this version of PyTorch was built for and the hardware CCs it supports:
- 7.5 which supports hardware CC >=7.5,<8.0
- 8.0 which supports hardware CC >=8.0,<9.0 except {8.7}
- 8.6 which supports hardware CC >=8.6,<9.0 except {8.7}
- 10.0 which supports hardware CC >=10.0,<11.0 except {10.1}
- 10.1 which supports hardware CC {10.1, 11.0}
Please follow the instructions at https://pytorch.org/get-started/locally/ to install a PyTorch release that supports one of these CUDA versions: 12.6, 12.8, 13.0
```

fixes: #164708

Reference: NVIDIA Compute Capability to Device Mapping: https://developer.nvidia.com/cuda-gpus

Pull Request resolved: https://github.com/pytorch/pytorch/pull/166185
Approved by: https://github.com/nWEIdia, https://github.com/atalman
